### PR TITLE
module2

### PR DESCRIPTION
**Pull request by pilinoga**
add module2

### CODE REVIEW

File: module2/controller/src/main/java/com/epam/esm/config/localization/Translator.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.context.support.ResourceBundleMessageSource;
+import org.springframework.stereotype.Component;
+import java.util.Locale;
+
+@Component
+public class Translator {
+
+    private static ResourceBundleMessageSource messageSource;
```
Comment (senator2202) [inline module2/controller/src/main/java/com/epam/esm/config/localization/Translator.java]: No need to make this field static, as Translator bean scope is singletone, by default

File: module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                            @RequestParam(required = false) String sort) {
+        List<Certificate> certificates = certificateService.search(tag, name, sort);
+        if(certificates.isEmpty()){
+            throw new NoSuchCertificateException();
+        }
+        return  certificates;
+    }
+
+    @PostMapping()
+    public ResponseEntity<?> saveCertificate(@RequestBody Certificate certificate){
```
Comment (senator2202) [inline module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java]: Why do you use "?", if you are returning concrete object, wrapped by ResponseEntity?

File: module2/service/src/main/java/com/epam/esm/service/CertificateServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            if (sort != null) {
+                if (sort.equalsIgnoreCase(ASC_SORT)) {
+                    return this.sort(byNameOrDescription,Comparator.comparing(Certificate::getName));
+                } else {
+                    return this.sort(byNameOrDescription,Comparator.comparing(Certificate::getName).reversed());
+                }
+            }
+            return byNameOrDescription;
+        }
+        return sort.equalsIgnoreCase(ASC_SORT)?
```
Comment (senator2202) [inline module2/service/src/main/java/com/epam/esm/service/CertificateServiceImpl.java]: What if sort == null ?

File: module2/model/src/main/java/com/epam/esm/dao/CertificateDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+            ps.setString(5, ZonedDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE));
+            ps.setString(6, ZonedDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE));
+            return ps;
+        }, keyHolder);
+        Long certificateId = keyHolder.getKey().longValue();
+        this.saveTagsFromCertificate(certificate, certificateId);
+    }
+
+    @Override
+    public void update(Certificate certificate, int id) {
```
Comment (senator2202) [inline module2/model/src/main/java/com/epam/esm/dao/CertificateDAOImpl.java]: I can see, that you are performing several operations in different tables here. But I can not see, where you use transactions

Comment (senator2202) [reply module2/model/src/main/java/com/epam/esm/dao/CertificateDAOImpl.java]: This issue must be fixed for all cases over the project where transactions must applied.

**Review by senator2202 — CHANGES_REQUESTED** [review]

File: module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java (hunk 2 above)
Comment (senator2202) [reply module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java:54]: Store intefaces in package root, there implementations in sub package impl:
image

**Comment by senator2202** [comment]
Store intefaces in package root, there implementations in sub package impl:
example: 
image

File: module2/model/src/main/java/com/epam/esm/model/Tag.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
+package com.epam.esm.model;
+
+import java.util.Objects;
+
+public class Tag {
+    private Long id;
+    private String name;
+
+    public Tag() {
```
Comment (senator2202) [inline module2/model/src/main/java/com/epam/esm/model/Tag.java:11]: Make sure that you use all constructors, that you have written

File: module2/controller/src/main/java/com/epam/esm/config/localization/Translator.java (hunk 1 above)
Comment (pilinoga) [reply module2/controller/src/main/java/com/epam/esm/config/localization/Translator.java]: > No need to make this field static, as Translator bean scope is singletone, by default

fix

File: module2/model/src/main/java/com/epam/esm/dao/CertificateDAOImpl.java (hunk 4 above)
Comment (pilinoga) [reply module2/model/src/main/java/com/epam/esm/dao/CertificateDAOImpl.java]: > This issue must be fixed for all cases over the project where transactions must applied.

add @Transactional

File: module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java (hunk 2 above)
Comment (pilinoga) [reply module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java]: > Why do you use "?", if you are returning concrete object, wrapped by ResponseEntity?

fix, use specific object type

File: module2/model/src/main/java/com/epam/esm/model/Tag.java (hunk 5 above)
Comment (pilinoga) [reply module2/model/src/main/java/com/epam/esm/model/Tag.java:11]: empty constructor is used by spring for tag bean initiliazation

File: module2/model/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    public List<Certificate> getAll(){
+        List<CertificateTag> list = certificateTagDAO.getAll();
+        List<Certificate> certificates = jdbcTemplate.query(GET_ALL_SQL,
+                new BeanPropertyRowMapper<>(Certificate.class));
+        return getCertificatesWithTags(list, certificates);
+    }
+
+    @Override
+    @Transactional
+    public void save(Certificate certificate) {
```
Comment (senator2202) [inline module2/model/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java]: Move multitable transactions on service layer

File: module2/model/src/main/java/com/epam/esm/model/Tag.java (hunk 5 above)
Comment (senator2202) [reply module2/model/src/main/java/com/epam/esm/model/Tag.java:11]: @pilinoga, is Tag managed by spring? Is it bean?

File: module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    public List<Certificate> getAllCertificates(){
+        return certificateService.getAll();
+    }
+
+    @GetMapping("/{id}")
+    public Certificate getCertificateById(@PathVariable int id){
+        return certificateService.getByID(id);
+    }
+
+    @GetMapping("/search")
```
Comment (senator2202) [inline module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java]: @pilinoga, No need for this "search"part in url. Just leave get mapping with parameters

File: module2/service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        return tagDAO.getAll();
+    }
+
+    @Override
+    public Tag getByID(int id) {
+        return tagDAO.getByID(id);
+    }
+
+    @Override
+    public Tag save(Tag tag) {
```
Comment (senator2202) [inline module2/service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java:31]: I can not see any validation. Am I allowed to add tag with empty name?

File: module2/service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    private final String ASC_SORT="ASC";
+
+    @Autowired
+    public CertificateServiceImpl(CertificateDAO certificateDAO) {
+        this.certificateDAO = certificateDAO;
+    }
+
+    @Override
+    public void save(Certificate certificate) {
+        certificateDAO.save(certificate);
```
Comment (senator2202) [inline module2/service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: Can not see any validation here

File: module2/model/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+import java.util.Map;
+import java.util.Set;
+
+@Component
+public class CertificateDAOImpl implements CertificateDAO {
+    private final JdbcTemplate jdbcTemplate;
+    private final TagDAO tagDAO;
+    private final CertificateTagDAO certificateTagDAO;
+
+    private final String GET_ALL_SQL = "SELECT * FROM certificate";
```
Comment (senator2202) [inline module2/model/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java]: We are declaring constants as **static final**

File: module2/model/pom.xml [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-jdbc</artifactId>
+            <version>5.3.13</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-tx</artifactId>
+            <version>5.3.13</version>
```
Comment (Lavazz) [inline module2/model/pom.xml]: could you please  write version values in properties tag

File: module2/model/src/main/java/com/epam/esm/dao/CertificateDAO.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package com.epam.esm.dao;
+
+import com.epam.esm.model.Certificate;
+import java.util.List;
+import java.util.Map;
+
+public interface CertificateDAO {
```
Comment (Lavazz) [inline module2/model/src/main/java/com/epam/esm/dao/CertificateDAO.java]: You can create common Dao interface with CRUD operations, and inherit from it

**Comment by senator2202** [comment]
I can not see this points from requirements:

3. Code should contain valuable comments where appropriate.
4. Public APIs should be documented (Javadoc).

File: module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+import javax.validation.Valid;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Class RestController represent api which allows to perform operations on tags.
```
Comment (senator2202) [inline module2/controller/src/main/java/com/epam/esm/controllers/CertificateController.java]: on tags?

File: module2/model/src/main/java/com/epam/esm/dao/DAO.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    List<T> getAll();
+
+    /**
+     * Method for getting an objects by id.
+     *
+     * @param id  used to find an object
+     * @return object
+     */
+    T getByID(int id);
```
Comment (senator2202) [inline module2/model/src/main/java/com/epam/esm/dao/DAO.java]: We are not getting, we are finding in database. Better name will be findByID. 
This should be applied to all similar cases

**Review by senator2202 — APPROVED** [review]

File: module2/service/src/main/java/com/epam/esm/service/Service.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
+package com.epam.esm.service;
+
+import java.util.List;
+
+public interface Service<T> {
```
Comment (senator2202) [inline module2/service/src/main/java/com/epam/esm/service/Service.java]: Not appropriate Name for Interface/Class. Name something like "BaseService",

**Review by senator2202 — CHANGES_REQUESTED** [review]

**Comment by senator2202** [comment]
@pilinoga, according to the task:
12. Service layer should be covered with unit tests not less than 80%.
Add unit tests on TagService as well.

**Review by senator2202 — APPROVED** [review]

**Pull request merged**